### PR TITLE
Add RLE decoder

### PR DESCRIPTION
**Pull request by mtdudek**
This PR adds a simple implementation of RLE decoder. 
It depends on:
- [ ] #1004 for intra RLE interfaces definitions.

Related: https://github.com/google/xls/issues/994

CC @proppy

### CODE REVIEW

**Comment by proppy** [comment]
Do want to land #974 first? or should we start reviewing now?

**Comment by tmichalak** [comment]
It seems we are close to merging #974, so we can start reviewing this soon.

**Comment by mtdudek** [comment]
I'm removing _draft_ label from this PR

**Comment by proppy** [comment]
can you rebase on top of #974?

File: xls/modules/rle/rle_interface.x [hunk 1]
```diff
@@ -0,0 +1,34 @@
+// Copyright 2023 The XLS Authors
```
Comment (proppy) [inline xls/modules/rle/rle_interface.x:1]: maybe rename this to `rle_common.x` to match the convention in https://github.com/google/xls/tree/main/xls/modules/aes?

File: xls/modules/rle/rle_interface.x [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+// This file implements a parametric RLE encoder
+//
+// The encoder uses Run Length Encoding (RLE) to compress the input stream
+// with symbols (data to compress). The output contains packets of data with
+// the symbols and the number of their occurrences in the input stream.
+
+// structure containing output data a Data preprocessor
+// and input to an RLE encoder
+pub struct PreData<SYMB_WIDTH: u32> {
```
Comment (proppy) [inline xls/modules/rle/rle_interface.x:23]: is there a better name for this? i.e: `DecData`?

File: xls/modules/rle/rle_dec.x [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        input_r: chan<DecInData<SYMBOL_WIDTH, COUNT_WIDTH>> in,
+        output_s: chan<DecOutData<SYMBOL_WIDTH>> out,
+    ) {(input_r, output_s)}
+
+    next (tok: token, state: RLEDecState<SYMBOL_WIDTH, COUNT_WIDTH>) {
+        let zero_input = DecInData { symbol: bits[SYMBOL_WIDTH]:0, count: bits[COUNT_WIDTH]:0, last: false };
+        let empty = state.count == bits[COUNT_WIDTH]:0;
+        let (input_tok, input) = recv_if(tok, input_r, empty, zero_input);
+        let (next_symbol, next_count, next_last) = if (empty) {
+            let t_count = input.count - bits[COUNT_WIDTH]:1;
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:87]: is `t_count` for `token_count`?

Comment (proppy) [inline xls/modules/rle/rle_dec.x:87]: seems like this could go after the `if` block, since we substract `1` in both case?

Comment (mtdudek) [reply xls/modules/rle/rle_dec.x:87]: it was for `temporary_count`, as it's just an intermediate value.

Comment (mtdudek) [reply xls/modules/rle/rle_dec.x:87]: sounds good to me, I'll change that

File: xls/modules/rle/rle_dec.x [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+    )}
+
+    config (
+        input_r: chan<DecInData<SYMBOL_WIDTH, COUNT_WIDTH>> in,
+        output_s: chan<DecOutData<SYMBOL_WIDTH>> out,
+    ) {(input_r, output_s)}
+
+    next (tok: token, state: RLEDecState<SYMBOL_WIDTH, COUNT_WIDTH>) {
+        let zero_input = DecInData { symbol: bits[SYMBOL_WIDTH]:0, count: bits[COUNT_WIDTH]:0, last: false };
+        let empty = state.count == bits[COUNT_WIDTH]:0;
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:84]: should we rename `empty` to `symbol_completed` to be more explicit?

Comment (mtdudek) [reply xls/modules/rle/rle_dec.x:84]: I think empty was good, but this will be more descriptive, so I'll change it

File: xls/modules/rle/rle_dec.x [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+        let (input_tok, input) = recv_if(tok, input_r, empty, zero_input);
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:85]: it might actually be simpler to put the recv inside the if?

File: xls/modules/rle/rle_dec.x [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+        let (next_symbol, next_count, next_last) = if (empty) {
+            let t_count = input.count - bits[COUNT_WIDTH]:1;
+            (input.symbol, t_count, input.last)
+        } else {
+            let t_count = state.count - bits[COUNT_WIDTH]:1;
+            (state.symbol, t_count, state.last)
+        };
+        let send_last = next_last & (next_count == bits[COUNT_WIDTH]:0);
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:93]: shouldn't that be `&&` here ?

File: xls/modules/rle/rle_dec.x [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+        let data_tok = send(input_tok, output_s, DecOutData {symbol: next_symbol, last: send_last});
+        RLEDecState {
+            symbol: next_symbol,
+            count: next_count,
+            last: next_last,
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:98]: should we still be setting `next_last` and `next_symbol` and `next_count` if we're done sending?

File: xls/modules/rle/rle_dec.x [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+const TEST_SYMBOL_WIDTH = u32:32;
+const TEST_COUNT_WIDTH  = u32:2;
+
+type TestSymbol = bits[TEST_SYMBOL_WIDTH];
+type TestCount  = bits[TEST_COUNT_WIDTH];
+type TestDecInData  = DecInData<TEST_SYMBOL_WIDTH, TEST_COUNT_WIDTH>;
+type TestDecOutData = DecOutData<TEST_SYMBOL_WIDTH>;
+
+
+//const TEST_SEND_LOOKUP = TestDecInData[10]:[
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:132]: remove commented block?

File: xls/modules/rle/rle_dec.x [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+        (terminator, dec_input_s, dec_output_r)
+    }
+
+    next(tok: token, state: ()) {
+
+        // Simple transaction without repeats
+        let trans_send: (TestSymbol, TestCount)[2] =
+            [(TestSymbol:0xA, TestCount:0x3), (TestSymbol:0xB, TestCount:0x1)];
+        let tok = for ((counter, (symbol_in, count_in)), tok): ((u32, (TestSymbol, TestCount)) , token) in enumerate(trans_send) {
+            let _last = counter == (array_size(trans_send) - u32:1);
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:165]: curious why you have `_` here?

File: xls/modules/rle/rle_dec.x [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+        ];
+        let tok = for ((counter, symbol_out), tok): ((u32, TestSymbol) , token) in enumerate(trans_recv) {
+            let _last = counter == (array_size(trans_recv) - u32:1);
+            let data_out = TestDecOutData{symbol: symbol_out, last: _last};
+            let (tok, dec_output) = recv(tok, dec_output_r);
+            let _ = trace_fmt!(
+                "Received {} transactions, symbol: 0x{:x}, last: {}",
+                counter, dec_output.symbol, dec_output.last
+            );
+            let _ = assert_eq(dec_output, data_out);
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:183]: should we assert `last` too?

File: xls/modules/rle/rle_dec.x (hunk 7 above)
Comment (mtdudek) [reply xls/modules/rle/rle_dec.x:98]: If we are done sending next_count is 0, so will stall at recv_if

File: xls/modules/rle/rle_dec.x [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+            (tok)
+        }(tok);
+
+        // Transaction with repeating symbols
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:187]: is the different thing we're testing here the symbol `0xC` being chunked across two packet? shouldn't that meant that we set `last` on the first one (as we don't overflow the only time it can happen is with two transactions)

File: xls/modules/rle/rle_dec.x (hunk 8 above)
Comment (mtdudek) [reply xls/modules/rle/rle_dec.x:132]: missed that, removed in next commit

File: xls/modules/rle/rle_dec.x (hunk 9 above)
Comment (mtdudek) [reply xls/modules/rle/rle_dec.x:165]: It' left over from when I had there name collision.
I can remove it, but it will lead to `last: last`

File: xls/modules/rle/rle_dec.x (hunk 3 above)
Comment (proppy) [reply xls/modules/rle/rle_dec.x:87]: oh I see, I think `count` is perfectly acceptable then.

File: xls/modules/rle/rle_dec.x (hunk 5 above)
Comment (mtdudek) [reply xls/modules/rle/rle_dec.x:85]: having `recv` outside `if` reduces returned tuple.
With `recv` in if, `tok` would have to be returned as well.

Comment (mtdudek) [reply xls/modules/rle/rle_dec.x:85]: so moving `recv` to `if` caused `segmentation fault` during DSLX to IR conversion

Comment (proppy) [reply xls/modules/rle/rle_dec.x]: did you already file an issue about this?

File: xls/modules/rle/rle_common.x [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file implements a parametric RLE encoder
+//
+// The encoder uses Run Length Encoding (RLE) to compress the input stream
+// with symbols (data to compress). The output contains packets of data with
+// the symbols and the number of their occurrences in the input stream.
+
+// structure containing output data a Data preprocessor
```
Comment (proppy) [inline xls/modules/rle/rle_common.x:21]: can you reword this comment?

File: xls/modules/rle/rle_dec.x [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+        let zero_input = DecInData { symbol: bits[SYMBOL_WIDTH]:0, count: bits[COUNT_WIDTH]:0, last: false };
+        let recv_next_symbol = (state.count == bits[COUNT_WIDTH]:0);
+        let (tok, input) = recv_if(tok, input_r, recv_next_symbol, zero_input);
+        let (next_symbol, count, next_last) = if (recv_next_symbol) {
+            (input.symbol, input.count, input.last)
+        } else {
+            (state.symbol, state.count, state.last)
+        };
+        let next_count = count - bits[COUNT_WIDTH]:1;
+        let send_last = next_last && (next_count == bits[COUNT_WIDTH]:0);
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:92]: should we capture `(next_count == bits[COUNT_WIDTH]:0)` condition with a name? i.e: `done_sending`

File: xls/modules/rle/rle_dec.x [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+        let data_tok = send(tok, output_s, DecOutData {symbol: next_symbol, last: send_last});
+        if (send_last) {
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:94]: I wonder if we could be more explicit here by flattening the three possible state (with a match statement):
https://google.github.io/xls/dslx_reference/#match-expression
```
  match (recv_next_symbol, done_sending, send_last) {
    (_, true, true) => zero!<RLEDecState>(),
    (true, _, _) => (input.symbol, input.count, input.last)
    (false, _, _) => (state.symbol, state.count, state.last)
  }
```
wdy?

File: xls/modules/rle/rle_dec.x (hunk 11 above)
Comment (proppy) [reply xls/modules/rle/rle_dec.x]: same commment here, is it valid in the protocol to heave multiple consecutive `(symbol, count)` packet with the same `symbol` without having `last` being set between them?

File: xls/modules/rle/rle_dec.x (hunk 7 above)
Comment (proppy) [reply xls/modules/rle/rle_dec.x]: clarified my comment with the `match` suggestion.

File: xls/modules/rle/rle_dec.x (hunk 5 above)
Comment (mtdudek) [reply xls/modules/rle/rle_dec.x]: Issue #1035

File: xls/modules/rle/rle_dec.x (hunk 14 above)
Comment (mtdudek) [reply xls/modules/rle/rle_dec.x:94]: I agree it will make this code cleaner.
I'll change it in next commit

File: xls/modules/rle/rle_dec.x [hunk 15]
```diff
@@ -0,0 +1,512 @@
...
+          count: next_count,
+          last: next_last,
+        }
+      }
+  }
+}
+
+
+// RLE decoder specialization for the codegen
+proc RLEDec32 {
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:108]: rename to match the state naming convention?

File: xls/modules/rle/rle_dec.x [hunk 16]
```diff
@@ -0,0 +1,512 @@
...
+struct RunLengthDecoderState<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+  // symbol to be repeated on output
+  symbol: bits[SYMBOL_WIDTH],
+  // count of symbols that has to be send
+  count: bits[COUNT_WIDTH],
+  // send last when repeat ends
+  last: bool,
+}
+ // RLE decoder implementation
+pub proc RLEDec<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:65]: rename to match the state naming convention?

File: xls/modules/rle/rle_dec.x (hunk 9 above)
Comment (proppy) [reply xls/modules/rle/rle_dec.x]: I think `last: last` is ok.

File: xls/modules/rle/rle_dec.x (hunk 16 above)
Comment (mtdudek) [reply xls/modules/rle/rle_dec.x]: Names updated

File: xls/modules/rle/rle_dec.x (hunk 9 above)
Comment (mtdudek) [reply xls/modules/rle/rle_dec.x]: I've changed all `_last` to `last`

File: xls/modules/rle/rle_dec.x (hunk 11 above)
Comment (mtdudek) [reply xls/modules/rle/rle_dec.x]: It's possible to receive multiple `(symbol, count)` pairs with the same `symbol` without `last`being set in the first one.
This can happen when either encoder's counter has overflown or some preceding stage has decided to send such packets. Decoder must be able to handle such transaction.

File: xls/modules/rle/rle_common.x (hunk 12 above)
Comment (mtdudek) [reply xls/modules/rle/rle_common.x]: I missed this comment before, now it's reworded.

File: xls/modules/rle/rle_dec.x (hunk 13 above)
Comment (mtdudek) [reply xls/modules/rle/rle_dec.x]: I've changed code to capture condition as requested

**Comment by mtdudek** [comment]
@proppy can you check if PR is good for merging, or some more work is required?

File: xls/modules/rle/rle_common.x [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+// It is also used as an output from RLE decoder.
+pub struct PlainData<SYMB_WIDTH: u32> {
+    symbol: bits[SYMB_WIDTH], // symbol
+    last: bool,               // flush RLE
+}
+
+// Structure contains compressed (symbol, counter) pairs.
+// Structure is used as an output from RLE encoder and
+// as an input to RLE decoder.
+pub struct EncData<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
```
[the remark this answers is not in the document]
Comment (mtdudek) [reply xls/modules/rle/rle_common.x:30]: Sure, I'll change it, and squash all commits into single one, wdyt? @proppy

File: xls/modules/rle/rle_dec.x [hunk 18]
```diff
@@ -0,0 +1,520 @@
...
+
+  next (tok: token, state: RunLengthDecoderState<SYMBOL_WIDTH, COUNT_WIDTH>) {
+    let zero_input = DecInData {
+      symbol: bits[SYMBOL_WIDTH]:0,
+      count: bits[COUNT_WIDTH]:0,
+      last: false
+    };
+    let recv_next_symbol = (state.count == bits[COUNT_WIDTH]:0);
+    let (tok, input) = recv_if(tok, input_r, recv_next_symbol, zero_input);
+    let (next_symbol, count, next_last) = if (recv_next_symbol) {
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:90]: I wonder if we could simply assign the `zero_input` so the current state and get rid of this if?

File: xls/modules/rle/rle_dec.x [hunk 19]
```diff
@@ -0,0 +1,520 @@
...
+    ()
+  }
+}
+
+// Tests
+
+const TEST_SYMBOL_WIDTH = u32:32;
+const TEST_COUNT_WIDTH  = u32:32;
+
+type TestCommonSymbol     = bits[TEST_SYMBOL_WIDTH];
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:137]: nit: should this be named after the test `SingleSymbol`?

Comment (proppy) [reply xls/modules/rle/rle_dec.x:137]: or simply drop the `Command` if it's shared between multiple tests.

File: xls/modules/rle/rle_dec.x [hunk 20]
```diff
@@ -0,0 +1,520 @@
...
+      );
+      assert_eq(dec_output, output);
+      (tok)
+    }(tok);
+    send(tok, terminator, true);
+    ()
+  }
+}
+
+// Check RLE decoder on longer transaction
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:451]: I think we can simply keep this test + the special cases (and drop the simpler one at the beginning), what do you think?

**Review by proppy — APPROVED** [review]
Looks good, just some nits around naming, simplification and testing.

File: xls/modules/rle/rle_dec.x (hunk 20 above)
Comment (mtdudek) [reply xls/modules/rle/rle_dec.x:451]: Sure thing, I'll see that 2 tests are redundant. I'll remove them in next update

Comment (mtdudek) [reply xls/modules/rle/rle_dec.x:322]: I hope it better now, wdyt @proppy?

File: xls/modules/rle/rle_dec.x (hunk 19 above)
Comment (mtdudek) [reply xls/modules/rle/rle_dec.x:132]: But it's `Common`, so maybe it can be left as it is, wdyt @proppy?

File: xls/modules/rle/rle_dec.x (hunk 18 above)
Comment (mtdudek) [reply xls/modules/rle/rle_dec.x]: Changed it

File: xls/modules/rle/rle_dec.x (hunk 19 above)
Comment (proppy) [reply xls/modules/rle/rle_dec.x:132]: yep, I understand it's being shared by the tests but I think the definition appearing at the beginning of the file is enough of a hint to convey that it's being shared (without having to prefix all the name by `Common`)

File: xls/modules/rle/rle_dec.x [hunk 21]
```diff
@@ -0,0 +1,391 @@
...
+#[test_proc]
+proc RunLengthDecoderrTransactionTest {
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:324]: s/Decoderr/Decoder/

File: xls/modules/rle/rle_dec.x (hunk 20 above)
Comment (proppy) [reply xls/modules/rle/rle_dec.x:322]: yep, I don't think we need `RunLengthDecoderSingleSymbolTest` and `RunLengthDecoderSinglePacketTest`, since we have the longer test.

I think test that cover corner cases (`last`, `overflow` packet patterns) would be more relevant.

File: xls/modules/rle/rle_dec.x (hunk 21 above)
Comment (mtdudek) [reply xls/modules/rle/rle_dec.x]: Fixed

File: xls/modules/rle/rle_dec.x (hunk 19 above)
Comment (mtdudek) [reply xls/modules/rle/rle_dec.x]: I've removed `Common` from naming.

**Comment by proppy** [comment]
I wonder if we should decouple this from https://github.com/google/xls/pull/974, in order to import both PRs concurrently (at the cost of a bit of duplication that we could resolve in the other PRs)

File: xls/modules/rle/rle_dec.x (hunk 20 above)
Comment (mtdudek) [reply xls/modules/rle/rle_dec.x:198]: I've removed these tests. I agree that they were redundant.

**Comment by proppy** [comment]
> mtdudek force-pushed the rle-decoder branch from d28c593 to 633e8da 

it's sometime challenging to review changes that are force-pushed, maybe we could send incremental commit and only force push at the end after rebasing once the review is complete?

**Comment by mtdudek** [comment]
My bad, sorry with these force pushes.

**Comment by mtdudek** [comment]
> I wonder if we should decouple this from #974, in order to import both PRs concurrently (at the cost of a bit of duplication that we could resolve in the other PRs)

I'll change first commit in the chain, so it will add interface file, but it will not change rle, and rle renaming will be removed.
This should be enough to decouple it from #974. I'll create new PR that only updates RLE and changes it's name. 
Is that ok with you  @proppy?

**Comment by proppy** [comment]
> I'll change first commit in the chain, so it will add interface file, but it will not change rle, and rle renaming will be removed.

That would work too, or we could simply duplicate the types and create a new PR that add `rle_common.x` and remove the duplication.

**Comment by mtdudek** [comment]
> > I'll change first commit in the chain, so it will add interface file, but it will not change rle, and rle renaming will be removed.
> 
> That would work too, or we could simply duplicate the types and create a new PR that add `rle_common.x` and remove the duplication.

I'll take this route. I'll have to force push it. @proppy please tell me when I can do it safely.

**Comment by proppy** [comment]
> I'll have to force push it

I think it's always fine to force push after all comments have been resolved and the PR is approved. It's just that it's useful to be able to review incremental commits before resolving pending comments.

**Comment by mtdudek** [comment]
Remove dependency on #974 @proppy

File: xls/modules/rle/rle_dec.x [hunk 22]
```diff
@@ -0,0 +1,274 @@
...
+      );
+      assert_eq(dec_output, output);
+      (tok)
+    }(tok);
+    send(tok, terminator, true);
+    ()
+  }
+}
+
+// Check RLE decoder on longer transaction
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:205]: can we put the general test case first?

Comment (proppy) [inline xls/modules/rle/rle_dec.x:205]: > "longer transaction"

It's not really "longer" since there is no more simple tests :)

File: xls/modules/rle/rle_dec.x [hunk 23]
```diff
@@ -0,0 +1,274 @@
...
+const TEST_SYMBOL_WIDTH = u32:32;
+const TEST_COUNT_WIDTH  = u32:32;
+
+type TestSymbol     = bits[TEST_SYMBOL_WIDTH];
+type TestCount      = bits[TEST_COUNT_WIDTH];
+type TestStimulus   = (TestSymbol, TestCount);
+type TestDecInData  = DecInData<TEST_SYMBOL_WIDTH, TEST_COUNT_WIDTH>;
+type TestDecOutData = DecOutData<TEST_SYMBOL_WIDTH>;
+
+// Check that RLE decoder will create 2 `last` output packets,
```
Comment (proppy) [inline xls/modules/rle/rle_dec.x:145]: q: are there other corner case we should test for?

Comment (mtdudek) [reply xls/modules/rle/rle_dec.x:145]: I think there are no more corner cases to be checked.

File: xls/modules/rle/rle_dec.x (hunk 22 above)
Comment (mtdudek) [reply xls/modules/rle/rle_dec.x:205]: Removed "longer" from the comment.

Comment (mtdudek) [reply xls/modules/rle/rle_dec.x]: Changed, so that simple tests is first

File: xls/modules/rle/rle_dec.x (hunk 23 above)
Comment (proppy) [reply xls/modules/rle/rle_dec.x]: what if send a packet with a `count` of `0`?

Comment (mtdudek) [reply xls/modules/rle/rle_dec.x]: I assumed that count is never 0, as encoder never emits pairs with count = 0.

Comment (proppy) [reply xls/modules/rle/rle_dec.x]: I see, that makes sense if we only care about testing valid input.

**Comment by proppy** [comment]
Feel free to rebase and squash.

**Comment by proppy** [comment]
@mtdudek actually, I just realized that if we want to land this and #974 concurrently we need to decouple the `BUILD` file too. Do you mind if we split this in separate directory and then unify them in a follow up PR (sorry for all the back and forth on this).

**Comment by mtdudek** [comment]
I'll move it to rle_dec folder for now.

File: xls/modules/rle/rle_dec.x (hunk 23 above)
Comment (grebe) [reply xls/modules/rle/rle_dec.x]: I think it is useful to treat count as being offset by 1 so count=0 means 1 symbol, count=1 means 2 symbols, etc. Disallowing count=0 is wasteful when count is narrow, e.g. in the case of count being 1 bit it will always be longer. If you encode count with an offset, it should probably be named something like `offset_count` or `count_minus_one` to avoid confusion.

Comment (grebe) [reply xls/modules/rle/rle_dec.x]: Regardless, as-is I think there should be a `fail!()` if count is 0, as otherwise it is interpreted as `2^COUNT_WIDTH` symbols.

Comment (mtdudek) [reply xls/modules/rle/rle_dec.x]: If you have nothing against it, for now I will add `fail!()` if count is `0`.

Comment (proppy) [reply xls/modules/rle/rle_dec.x]: @mtdudek do you think we should file an issue to capture @grebe's offset by 1 suggestion? I think it would be worth discussing it.

File: xls/modules/rle_dec/rle_dec.x [hunk 24]
```diff
@@ -94,7 +94,11 @@ pub proc RunLengthDecoder<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
     };
     let recv_next_symbol = (state.count == bits[COUNT_WIDTH]:0);
     let (tok, input) = recv_if(tok, input_r, recv_next_symbol, state_input);
-    let next_count = input.count - bits[COUNT_WIDTH]:1;
+    let next_count = if input.count == bits[COUNT_WIDTH]:0 {
+        fail!("Count_eq_0", input.count)
```
Comment (proppy) [inline xls/modules/rle_dec/rle_dec.x:98]: nit: rename to `invalid_count_0`?

File: xls/modules/rle_dec/rle_dec.x [hunk 25]
```diff
@@ -94,7 +94,11 @@ pub proc RunLengthDecoder<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
     };
     let recv_next_symbol = (state.count == bits[COUNT_WIDTH]:0);
     let (tok, input) = recv_if(tok, input_r, recv_next_symbol, state_input);
-    let next_count = input.count - bits[COUNT_WIDTH]:1;
+    let next_count = if input.count == bits[COUNT_WIDTH]:0 {
```
Comment (proppy) [inline xls/modules/rle_dec/rle_dec.x:97]: q: I don't think we have a way to test for `fail!`, should we file a feature request for `assert_fail!`?

**Review by proppy — APPROVED** [review]
just some optional nits, can you rebase and and squash?

File: xls/modules/rle/rle_dec.x (hunk 23 above)
Comment (proppy) [reply xls/modules/rle/rle_dec.x]: filed https://github.com/google/xls/issues/1070

File: xls/modules/rle_dec/rle_dec.x (hunk 24 above)
Comment (mtdudek) [reply xls/modules/rle_dec/rle_dec.x]: Updated @proppy

**Comment by mtdudek** [comment]
Squashed and rebased
